### DISCUSSION TITLE
NUnit 4.0.0

### DISCUSSION THREAD

**Discussion by OsirisTerje** [nunit/nunit#4563] [Releases]
See [release notes](https://docs.nunit.org/articles/nunit/release-notes/framework.html#nunit-400---november-26-2023)
and
See [migration guide](https://docs.nunit.org/articles/nunit/release-notes/Nunit4.0-MigrationGuide.html)

<hr /><em>This discussion was created from the release <a href='https://github.com/nunit/nunit/releases/tag/'>NUnit 4.0.0</a>.</em>

**Comment by Sputnik24** [comment]
I migrated my testcase to NUnit 4.0.0 and my own console agent to NUnit 4.0.0 and NUnit.Engine 3.15.5 (was using 3.16.4 before).

Unfortunately, running the test using the local TestRunner does not work anymore.

```csharp
TestEngine.WorkDirectory = Path.GetDirectoryName(testFile);
TestPackage = new TestPackage(testFile);
TestRunner = TestEngine.GetRunner(TestPackage);
TestNodes = TestRunner.Explore(TestFilter.Empty);
```

Gives the following:
NUnit.Engine.NUnitEngineException: An exception occurred in the driver while loading tests.
---> System.IO.FileNotFoundException: Could not load file or assembly '_TestFileAssembly_, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. Das System kann die angegebene Datei nicht finden.

Do I do something wrong?

**Reply by OsirisTerje** [reply]
There are differences between the 3.16.X and 3.15.X series of engines.  Can you try to see if it works when you ONLY change that, that is keep NUnit on v3, and just change the engine.  If that still fails it is due to differences between these two engine series and not related to NUnit 4.

**Reply by Sputnik24** [reply]
Using NUnit 3.14.0 and Engine 3.15.5, I receive the same error and I found an issue on that in the Engine: https://github.com/nunit/nunit-console/issues/1180 but this does not help. I questioned it here because release notes say that NUnit 4 only works with Engine 3.15.3 or greater.

Btw: I'm using .NET 8

**Reply by OsirisTerje** [reply]
We have not done anything with the 3.16.X series to make it cope with NUnit 4.  We have not tested it either, so *if* it works, then that is due to earlier work.  I do see that there was added some early support for NUnit 4, but afaic see this was for the engine v4.

So if your agent works with the 3.16.X engine, can you try that with NUnit 4?

**Reply by Sputnik24** [reply]
Engine 3.16.X is not working with NUnit v4. It says

System.Exception: No suitable tests found in '_testfile_.dll'."
Either assembly contains no tests or proper test driver has not been found.

**Reply by OsirisTerje** [reply]
Ok, thanks for verifying that. 

The fix in 3.16 for the driver issue must then be found and ported to 3.15.

IIRC:  There has been some discussions around this issue earlier.  There might be some suggestion in some of these.  I'll try to have a look around, also feel free to search through the issues for this.

There is also an alternative in locating the NUnit 4 fix in 3.15, I remember it was fairly simple, we could then port that to 3.16.

**Reply by Sputnik24** [reply]
Hi @OsirisTerje is there an update on this topic?

**Reply by OsirisTerje** [reply]
Sorry, no progress on it.  We're out of resources in the console/engine project, so it will be slow.

**Reply by stevenaw** [reply]
@Sputnik24 I noticed you mentioned you're using .NET8. I think .NET8 support was added in version 3.17 of the console runner. Have you tried to see if it works for you?

**Reply by stevenaw** [reply]
In terms of changes in 3.16 which aren't in 3.15/3.17, there were a few PRs which affected the `/Drivers` directory (`/nunit.engine.core/Drivers`) which aren't in 3.17:
- https://github.com/nunit/nunit-console/pull/1249
- https://github.com/nunit/nunit-console/pull/1250
- https://github.com/nunit/nunit-console/pull/1267

The first only looks to improve an error message but the second and third touch some of the assembly loading logic. If I'm reading it right the third may also remove a call to `var assemblyPath = _resolver.ResolveAssemblyToPath(name);` as part of some refactoring to the `CustomAssemblyLoadContext`

@Sputnik24  If 3.17 does not work for you then I would suggest to make an issue to start a discussion about porting those changes (and identifying any others), however we would want to be careful of course to not re-introduce the issues from the 3.16 branch again.

**Reply by stevenaw** [reply]
Oh. Perhaps PR https://github.com/nunit/nunit-console/pull/1376/files could help here too

**Comment by SimonCropp** [comment]
is there going to be a NUnit4TestAdapter?

**Reply by OsirisTerje** [reply]
We have an unfortunate naming issue there.  We cant rename the package, so we're stuck with NUnit3Testadapter- at least until Microsoft/Nuget figures out a way to rename packages but keeping the identity.  If that happens,  we can rename it back to NUnitTestAdapter.
Version 4.5 of the NUnit3TestAdapter IS the de-facto version 4 of the adapter, and runs all versions of NUnit 3 and 4.

**Reply by OsirisTerje** [reply]
Also worth mentioning, moving forward we will follow Semver,  so the major numbers of both will increase for breaking changes.  The naming of the packages then have to remain the same.  The idea of having versions in the names was very unfortunate,

**Reply by OsirisTerje** [reply]
An alternative could be  to create a new package of the same with e.g. the name NUnitAdapter.  But, it will be  a major breaking change for very many people.  So not sure if we can do this.

**Comment by seesharks** [comment]
A bit of a crossover issue, but I've upgraded to 4.0.0 (including using the analyzer to automatically port to the constraint model -- good stuff by the way -- it worked like a dream!). Now my ReSharper test runner comes up empty when scanning projects for tests. It finds none whatsoever.

The built in VS test runner with NUnit3TestAdapter installed does find and successfully run all tests. It's just ReSharper that has an issue.

Does anyone know if ReSharper is compatible with the new version?

This post seems relevant, but I am well ahead of the fix version it indicates: https://github.com/nunit/nunit3-vs-adapter/issues/846

**Comment by OsirisTerje** [comment]
You're absolutely 100% correct!  It doesn't work on version 2023.2, but that has just recently been fixed in 2023.3 which is in EAP now, hopefully soon released.

PS: The issue you link to is for the adapter, which Resharper is ignoring actually, but that issue was fixed.  The issue now is that Rider/REsharper doesnt support NUnit 4, because they have used an outdated engine as part of their solution.  Now they have fixed that too, so either try the EAP or wait for the release of 2023.3

**Reply by seesharks** [reply]
Got it, thanks for the quick response!

**Comment by PzYon** [comment]
I had the same issue. Rider 2023.3 has been released and now it works.

**Reply by jdebache** [reply]
It's a Rider issue I guess, but still not really working for me. I usually have to invalidate caches + restart + `git clean -xdf` to get the tests to run correctly.
